File: services/incident-service/src/main/java/io/aegisops/incident/messaging/IncidentEvent.java
package io.aegisops.incident.messaging;

import io.aegisops.incident.domain.Incident;
import io.aegisops.incident.domain.IncidentSeverity;
import io.aegisops.incident.domain.IncidentStatus;

import java.time.Instant;
import java.util.UUID;

public record IncidentEvent(
        UUID eventId,
        IncidentEventType eventType,
        int eventVersion,
        Instant occurredAt,
        UUID incidentId,
        String serviceName,
        IncidentSeverity severity,
        IncidentStatus status,
        IncidentStatus previousStatus
) {

    private static final int CURRENT_VERSION = 1;

    public static IncidentEvent created(Incident incident) {
        return new IncidentEvent(
                UUID.randomUUID(),
                IncidentEventType.INCIDENT_CREATED,
                CURRENT_VERSION,
                incident.createdAt(),
                incident.id(),
                incident.serviceName(),
                incident.severity(),
                incident.status(),
                null
        );
    }

    public static IncidentEvent statusChanged(
            Incident incident,
            IncidentStatus previousStatus
    ) {
        return new IncidentEvent(
                UUID.randomUUID(),
                IncidentEventType.INCIDENT_STATUS_CHANGED,
                CURRENT_VERSION,
                incident.updatedAt(),
                incident.id(),
                incident.serviceName(),
                incident.severity(),
                incident.status(),
                previousStatus
        );
    }
}

File: services/realtime-gateway/src/kafka/incident-event.ts
import { z } from "zod";

export const incidentEventSchema = z
  .object({
    eventId: z.string().uuid(),
    eventType: z.enum([
      "INCIDENT_CREATED",
      "INCIDENT_STATUS_CHANGED"
    ]),
    eventVersion: z.literal(1),
    occurredAt: z.string().datetime({
      offset: true
    }),
    incidentId: z.string().uuid(),
    serviceName: z.string().min(1),
    severity: z.string().min(1),
    status: z.string().min(1),
    previousStatus: z.string().min(1).nullable()
  })
  .strict();

export type IncidentEvent = z.infer<
  typeof incidentEventSchema
>;

export function parseIncidentEvent(
  value: string
): IncidentEvent {
  let parsed: unknown;

  try {
    parsed = JSON.parse(value);
  } catch {
    throw new Error("Kafka message contains invalid JSON");
  }

  return incidentEventSchema.parse(parsed);
}